module/gui/Collector.py

```diff
@@ -66,79 +92,512 @@ def translateStatus(self, string):
         """
         return translatedStatusMap[string]
     
+    def getSelectionInfo(self):
+        """
+            called from MainWindow
+        """
+        func_start_time = self.time_msec()
+        QMutexLocker(self.mutex)
+        packsCnt = linksCnt = 0
+        downloading = False
+        for si in self.view.selectionModel().selectedIndexes():
+            if si.column() != 0:
+                continue
+            item = si.internalPointer()
+            if isinstance(item, Package):
+                packsCnt += 1
+                if not downloading:
+                    for child in item.children:
+                        if child.data["downloading"]:
+                            downloading = True
+                            break
+            elif isinstance(item, Link):
+                linksCnt += 1
+                if not downloading:
+                    if item.data["downloading"]:
+                        downloading = True
+            else:
+                raise TypeError("Unknown item instance")
+        allPacksCnt = len(self._data)
+        self.log.debug8("%s.getSelectionInfo took %dms, packsCnt:%d linksCnt:%d downloading:%s allPacksCnt:%d" % (self.cname, self.time_msec() - func_start_time, packsCnt, linksCnt, downloading, allPacksCnt))
+        return (packsCnt, linksCnt, downloading, allPacksCnt)
+    
+    def getSelection(self, deselect, linksOfPacks):
+        """
+            called from main
+            deselect:     deselect the returned items
+            linksOfPacks: if deselect, for the links, only deselect those which belong to a selected package
+        """
+        func_start_time = self.time_msec()
+        QMutexLocker(self.mutex)
+        selection = []          # (pid, fid, isPack)
+        smodel = self.view.selectionModel()
+        for index in smodel.selectedRows(0):
+            item = index.internalPointer()
+            if isinstance(item, Package):
+                selection.append((item.id, None, True))
+            elif isinstance(item, Link):
+                selection.append((item.data["package"], item.id, False))
+            else:
+                raise TypeError("%s: Unknown item instance" % self.cname)
+            if deselect and not linksOfPacks:
+                smodel.select(index, QItemSelectionModel.Deselect | QItemSelectionModel.Rows)
+        if deselect and linksOfPacks:
+            for packindex in smodel.selectedRows(0):
+                pack = packindex.internalPointer()
+                if isinstance(pack, Package):
+                    smodel.select(packindex, QItemSelectionModel.Deselect | QItemSelectionModel.Rows)
+                    for linkindex in smodel.selectedRows(0):
+                        link = linkindex.internalPointer()
+                        if isinstance(link, Link) and link.data["package"] == pack.id:
+                            smodel.select(linkindex, QItemSelectionModel.Deselect | QItemSelectionModel.Rows)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['QItemSelectionModel' may be undefined, or defined from star imports: PyQt4.QtCore, PyQt4.QtGui](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=567903)